In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/financial_data.csv')
df

Company  Fiscal Year  Total Revenue($Million)  Net Income($Million)  \
0  Microsoft         2021                   168088                 61271   
1  Microsoft         2022                   198270                 72738   
2  Microsoft         2023                   211915                 72361   
3      Tesla         2021                    53823                  5644   
4      Tesla         2022                    81462                 12587   
5      Tesla         2023                    96773                 14974   
6      Apple         2021                   365817                 94680   
7      Apple         2022                   394328                 99803   
8      Apple         2023                   383285                 96995   

   Total Assets($Million)  Total Liabilities($Million)  \
0                  333779                       191791   
1                  364840                       198298   
2                  411976                       205753   
3                   62131                        30548   
4                   82338                        36440   
5                  106618                        43009   
6                  351002                       287912   
7                  352755                       302083   
8                  352583                       290437   

   Cash Flow from Operating Activities($Million)  
0                                          76740  
1                                          89035  
2                                          87582  
3                                          11497  
4                                          14724  
5                                          13256  
6                                         104038  
7                                         122151  
8                                         110543

In [3]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue($Million)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income($Million)'].pct_change() * 100

In [4]:
df

Company  Fiscal Year  Total Revenue($Million)  Net Income($Million)  \
0  Microsoft         2021                   168088                 61271   
1  Microsoft         2022                   198270                 72738   
2  Microsoft         2023                   211915                 72361   
3      Tesla         2021                    53823                  5644   
4      Tesla         2022                    81462                 12587   
5      Tesla         2023                    96773                 14974   
6      Apple         2021                   365817                 94680   
7      Apple         2022                   394328                 99803   
8      Apple         2023                   383285                 96995   

   Total Assets($Million)  Total Liabilities($Million)  \
0                  333779                       191791   
1                  364840                       198298   
2                  411976                       205753   
3                   62131                        30548   
4                   82338                        36440   
5                  106618                        43009   
6                  351002                       287912   
7                  352755                       302083   
8                  352583                       290437   

   Cash Flow from Operating Activities($Million)  Revenue Growth (%)  \
0                                          76740                 NaN   
1                                          89035           17.956071   
2                                          87582            6.882030   
3                                          11497                 NaN   
4                                          14724           51.351653   
5                                          13256           18.795267   
6                                         104038                 NaN   
7                                         122151            7.793788   
8                                         110543           -2.800461   

   Net Income Growth (%)  
0                    NaN  
1              18.715216  
2              -0.518299  
3                    NaN  
4             123.015592  
5              18.964010  
6                    NaN  
7               5.410858  
8              -2.813543

In [5]:
# Fill NA values that result from pct_change calculations with 0 or an appropriate value
df.fillna(0, inplace=True)

In [6]:
df

Company  Fiscal Year  Total Revenue($Million)  Net Income($Million)  \
0  Microsoft         2021                   168088                 61271   
1  Microsoft         2022                   198270                 72738   
2  Microsoft         2023                   211915                 72361   
3      Tesla         2021                    53823                  5644   
4      Tesla         2022                    81462                 12587   
5      Tesla         2023                    96773                 14974   
6      Apple         2021                   365817                 94680   
7      Apple         2022                   394328                 99803   
8      Apple         2023                   383285                 96995   

   Total Assets($Million)  Total Liabilities($Million)  \
0                  333779                       191791   
1                  364840                       198298   
2                  411976                       205753   
3                   62131                        30548   
4                   82338                        36440   
5                  106618                        43009   
6                  351002                       287912   
7                  352755                       302083   
8                  352583                       290437   

   Cash Flow from Operating Activities($Million)  Revenue Growth (%)  \
0                                          76740            0.000000   
1                                          89035           17.956071   
2                                          87582            6.882030   
3                                          11497            0.000000   
4                                          14724           51.351653   
5                                          13256           18.795267   
6                                         104038            0.000000   
7                                         122151            7.793788   
8                                         110543           -2.800461   

   Net Income Growth (%)  
0               0.000000  
1              18.715216  
2              -0.518299  
3               0.000000  
4             123.015592  
5              18.964010  
6               0.000000  
7               5.410858  
8              -2.813543

## Data Exploration and Analysis

In [7]:
# Average revenue growth for each company
avg_revenue_growth = df.groupby('Company')['Revenue Growth (%)'].mean()
print("Average Revenue Growth (%) by Company:")
print(avg_revenue_growth)

Average Revenue Growth (%) by Company:
Company
Apple         1.664442
Microsoft     8.279367
Tesla        23.382306
Name: Revenue Growth (%), dtype: float64


**Apple: 1.7%**<br>
Apple’s average revenue growth is positive, indicating a modest increase in revenue over the analyzed period.

<b>Microsoft: 8.3%</b><br>
Microsoft’s average revenue growth shows a significant increase, suggesting strong revenue growth over the period.

<b>Tesla: 23.4%</b><br>
Tesla’s average revenue growth is the highest, reflecting substantial revenue growth over the period.

In [8]:
# Average net income growth for each company
avg_net_income_growth = df.groupby('Company')['Net Income Growth (%)'].mean()
print("\nAverage Net Income Growth (%) by Company:")
print(avg_net_income_growth)


Average Net Income Growth (%) by Company:
Company
Apple         0.865772
Microsoft     6.065639
Tesla        47.326534
Name: Net Income Growth (%), dtype: float64


<b>Apple: 0.9%</b><br>
Apple's average net income growth is relatively modest, reflecting a slight increase in profitability over the past three years.

<b>Microsoft: 6.1%</b><br>
Microsoft's average net income growth is more substantial, indicating effective profitability growth during this period.

<b>Tesla: 47.3%</b><br>
Tesla’s average net income growth is the highest by a significant margin, showing exceptional improvement in profitability over the past three years.

In [9]:
# Total revenue by company
total_revenue_by_company = df.groupby('Company')['Total Revenue($Million)'].sum()
print("\nTotal Revenue by Company ($Million):")
print(total_revenue_by_company)


Total Revenue by Company ($Million):
Company
Apple        1143430
Microsoft     578273
Tesla         232058
Name: Total Revenue($Million), dtype: int64


<b>Apple: $1,143,430 million</b><br>
Apple has the highest total revenue, indicating its significant market presence and substantial revenue generation over the three-year period.

<b>Microsoft: $578,273 million</b><br>
Microsoft’s total revenue is substantial but less than Apple’s, reflecting its strong revenue performance but not as high as Apple’s.

<b>Tesla: $232,058 million</b><br>
Tesla has the lowest total revenue among the three, which is still considerable but lower compared to Apple and Microsoft.